r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs

```diff
@@ -345,22 +491,145 @@ for AffineGadget<P, ConstraintF, F>
         ))
     }
 
+    /// Hopwood's optimized fixed-length variable base exponentiation.
+    /// The scalar bits (in little endian representation) must be free of leading zeroes, 
+    /// and at most as long as the scalar field modulus.
+    /// 
+    /// The algorithm uses the same symmetrization technique as the signed lookup-table fixed
+    /// base scalar multiplication, and relies on the optimized double_and_add_internal in 
+    /// non-safe setting, as affine arithmetic exceptions are excluded from the presumptions on
+    /// the scalar bits (and an extra treatment of &self being trivial).
+    ///
+    /// Given an exponent of length n+1, scalar = 2^n + k with len(k) <= n, compute  
+    /// [bits] self = [2^n + k] T
+    ///
+    /// Acc := [3] T
+    /// for i from n-2 down to 0 {
+    ///     Q := k[i+1] ? T : −T
+    ///     Acc := (Acc + Q) + Acc
+    /// }
+    /// return (k[0] = 0) ? (Acc - T) : Acc
+    /// 
+    /// See https://github.com/zcash/zcash/issues/3924 for a detailed explanation.
+    /// Implementation adapted from https://github.com/ebfull/halo/blob/master/src/gadgets/ecc.rs#L1762
+    fn mul_bits<'a, CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        bits: impl Iterator<Item = &'a Boolean>,
+    ) -> Result<Self, SynthesisError> {
+
+        // TODO: we must assert that bits.length() <= ScalarField::MODULUS.length(), 
+        // otherwise the algorithm is not secure.
```

Comment:
Alternatively, we may even want to restrict that the scalar is below the scalar field modulus.